cmd/snap-confine/mount-support-wsl2-gpu.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+#include "mount-support-wsl2-gpu.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <glob.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <stdint.h>
+#include <unistd.h>
+/* POSIX version of basename() and dirname() */
+#include <libgen.h>
+
+#include "../libsnap-confine-private/classic.h"
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/utils.h"
+
+// note: if the parent dir changes to something other than
+// the current /var/lib/snapd/lib then sc_mkdir_and_mount_and_bind
+// and sc_mkdir_and_mount_and_bind need updating.
+#define SC_LIB "/var/lib/snapd/lib"
+#define SC_WSL_GPU_DIR   SC_LIB "/wsl"
```

Comment:
could those libraries be actually mounted under `/var/lib/snapd/lib/gl` which is already included in SNAP_LIBRARY_PATH?

Reply:
no, because `/var/lib/snapd/lib/gl` is a mountpoint OR a symlink dumping ground depending upon which distro snapd is running on (e.g. IIRC arch needs symlinks instead of mountpoint). Because it might be a host mountpoint, not a directory containing symlinks, then we can't also mount the wsl directory there assuming the nvidia driver isn't already mounted, and adding symlinks there for wsl libraries will potentially pollute the host if the nvidia libraries are mounted already.

While there is defensive coding to only mount the nvidia libs and separately the wsl libs there is a potential that both may be present on a system at the same time (user error?) which will cause issues when sharing the directory.